### PR TITLE
ADR 17: Enable container autodiscovery to scan images from private repos

### PR DESCRIPTION
**Pull request by the-simmon**
First draft, mostly brain storming, feel free to commit to this branch :) (this time with correct adr number)

### CODE REVIEW

**Review by SebieF — COMMENTED** [review]
Very nice work, thank you :) Only few minor comments from my side.

It might be worth to include the trivy operator and k8s scan functionality as an additional possibility? Other than that, I really like the sidecar proposal and that one is already very sophisticated thanks to your example!

**Comment by the-simmon** [comment]
> Very nice work, thank you :) Only few minor comments from my side.
> 
> It might be worth to include the trivy operator and k8s scan functionality as an additional possibility? Other than that, I really like the sidecar proposal and that one is already very sophisticated thanks to your example!

I don think it is possible to incorporate the trivy operator into the scb as it functions differently. I looked at the trivy docs and added the idea to use ownerReferences to delete the secret after the scan is done to this adr.

File: docs/adr/adr_0017.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
+<!--
+SPDX-FileCopyrightText: the secureCodeBox authors
+
+SPDX-License-Identifier: Apache-2.0
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:4]:
```suggestion
SPDX-License-Identifier: Apache-2.0
title: "ADR-0017: Enable Container Autodiscovery To Scan Container Images From Private Registries
sidebar_label: "ADR-0017"
```

File: docs/adr/adr_0017.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+-->
+
+# ADR-0017: Enable container autodiscovery to scan container images from private registries
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:7]:
```suggestion
# ADR-0017: Enable Container Autodiscovery To Scan Container Images From Private Registries
```

File: docs/adr/adr_0017.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+<!--
+  Use one of the ADR status parameter based on status
+  Please add a cross reference link to the new ADR on 'superseded' ADR.
+  e.g.: {adr_suposed_by} <<ADR-0000>>
+-->
+| <!-- -->       | <!-- --> |
+|----------------|----------|
+| **Status**:    | DRAFT [ADR-0017](ADR-0017.md) |
+| **Date**:      | 2022-09-13 |
+| **Author(s)**: | Simon Hülkenberg <simon.huelkenberg@iteratec.com> |
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:18]:
```suggestion
| **Author(s)**: | Simon Hülkenberg <simon.huelkenberg@iteratec.com>, Sven Strittmatter <sven.strittmatter@iteratec.com> |
```

File: docs/adr/adr_0017.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+## Context
+
+The secureCodeBox offers an `autodiscovery` feature which automatically creates scheduled scans when new `services` or `containers` (in pods) are created. The container autodiscovery will start a trivy container scan for every distinct container in a given namespace. Currently it is not possible to scan container images that require `imagePullSecrets` to download the image. The container autodiscovery will create a scheduled scan but the scan itself will fail. Because of that limitation it is necessary to make the container autodiscovery aware of secrets to enable trivy to download the container images.
+
+### Problem regarding imagePullSecretes
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:24]:
```suggestion
### Problem Regarding imagePullSecretes

```

File: docs/adr/adr_0017.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+It is possible to configure multiple `imagePullSecrets` in a pods yaml file as seen in this [Stackoverflow](https://stackoverflow.com/questions/67073039/multiple-imagepullsecrets-in-a-kubernetes-deployment-for-the-same-url) post. Kubernetes will then try all available credentials until one succeeds. The container autodiscovery needs to do this too because trivy does not support multiple `imagePullSecrets` at once. Additionally one can only pass trivy `imagePullSecrets` as an environment variable whose name depends on the location of the image (see trivy [docs](https://aquasecurity.github.io/trivy/v0.17.0/private-registries/)). For google cloud credentials one as to provide a json file path as an environment variable for [example](https://aquasecurity.github.io/trivy/v0.17.0/private-registries/gcr/), but for dockerhub username and password as environment variable is sufficient. So trivy configuration changes quite significantly with different platforms.  
+Another problem is the fact that `imagePullSecrets` might change over time. It must be ensured that trivy has access to the newest `imagePullSecrets` when the scan starts.  
+
+### Possible solutions
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:28]:
```suggestion
### Possible Solutions

```

File: docs/adr/adr_0017.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+In general the easiest solution to provide trivy with the required credentials are kubernetes secretes that are mounted as environment variables. The following solutions use this idea although the secrets are created through different approaches.
+#### Using an initContainer and a sidecar
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:30]:
```suggestion
#### Using an initContainer and a Sidecar

```

File: docs/adr/adr_0017.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+This option does not alter the behavior of the container autodiscovery at all. All additional logic is moved to an `initContainer` and a `sidecar` container.  
+The `initContainer` reads the `imagePullSecrets` of the pod that houses the container to be scanned. The `initContainer` must determine which secret is the correct one for the scanned container image. It will then create a temporary secret following a predefined naming scheme (technically the trivy pod will be created while the secret that is mounted as an environment variable doesn't exit yet, thus the temporary secret name must be predefined). After the `initContainer` created the secret the trivy container and the `sidecar` will be started. The sidecar will continuously check if the main scan container exited (similar to the lurker) and then delete the temporary secret that was used to pass the credentials to trivy to clean up after the scan to avoid cluttering of the namespace.   
+Pros:
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:33]:
```suggestion

##### Pros

```

File: docs/adr/adr_0017.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+#### Using an initContainer and a sidecar
+This option does not alter the behavior of the container autodiscovery at all. All additional logic is moved to an `initContainer` and a `sidecar` container.  
+The `initContainer` reads the `imagePullSecrets` of the pod that houses the container to be scanned. The `initContainer` must determine which secret is the correct one for the scanned container image. It will then create a temporary secret following a predefined naming scheme (technically the trivy pod will be created while the secret that is mounted as an environment variable doesn't exit yet, thus the temporary secret name must be predefined). After the `initContainer` created the secret the trivy container and the `sidecar` will be started. The sidecar will continuously check if the main scan container exited (similar to the lurker) and then delete the temporary secret that was used to pass the credentials to trivy to clean up after the scan to avoid cluttering of the namespace.   
+Pros: 
+* The autodiscovery logic will not change.
+* No secrets that clutter the namespace
+
+Cons:
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:37]:
```suggestion
##### Cons:

```

File: docs/adr/adr_0017.md [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+Pros: 
+* The autodiscovery logic will not change.
+* No secrets that clutter the namespace
+
+Cons: 
+* By default a pod does not have sufficient permissions to create/delete secrets, so the pod needs to be assigned a `serviceAccount` with those permissions.   
+* Logic split up over multiple containers, solution a bit hacky
+* Edge Case: Changing the host platform of the container (example: AWS -> GCP) would break the scan because the required environment variables for trivy must change (as described above). The parameters of the scheduled scan must be changed but the existing autodiscovery operator would lack the logic to do this (no read permissions on secrets so it would notice the difference).
+<details><summary> Proof of concept</summary>
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:41]:
```suggestion

<details>
<summary> Proof of concept</summary>

```

File: docs/adr/adr_0017.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        - pacman -Sy && pacman -S --noconfirm kubectl && while [ ! -f /shared-volume/shutdown ]; do echo shutdownfile not found && sleep 2; done; kubectl delete secrets some-image-pull-secret
+        image: archlinux
+        name: secret-deletion-on-stop-sidecar
+        volumeMounts:
+        - name: shared-volume
+          mountPath: /shared-volume
+```
+</details>
+
+#### Using an initContainer and use ownerReferences for the created secretes
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:127]:
```suggestion
#### Using an initContainer and Use ownerReferences for the Created Secretes

```

File: docs/adr/adr_0017.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+Using the same initContainer logic described above but setting the ownerReference of the created secrets to the job/pod of the scan would result in similar behavior without the need of a sidecar.
+
+Pros:
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:130]:
```suggestion
##### Pros

```

File: docs/adr/adr_0017.md [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+* No need for a sidecar while keeping the namespace clean of temporary secrets
+
+#### Using a separate operator
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:133]:
```suggestion
#### Using a Separate Operator

```

File: docs/adr/adr_0017.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+Pros:
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:136]:
```suggestion
##### Pros

```

File: docs/adr/adr_0017.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+* Existing autodiscovery does need permissions to read secrets
+* It would be possible to only install the existing autodiscovery in case one doesn't like that the secret operator needs read permissions on secrets (Container Autodiscovery and new secret operator could be separate helm charts).
+* One could alter the roles of the service account of the new secret operator to only enable it in certain namespaces while using the existing autodiscovery in every namespace (both could use different serviceAccounts).
+
+Cons:
```
Comment (Weltraumschaf) [inline docs/adr/adr_0017.md:141]:
```suggestion
##### Cons

```

**Review by Weltraumschaf — CHANGES_REQUESTED** [review]
LGTM. Did some suggestions.

**Review by Weltraumschaf — APPROVED** [review]

**Comment by Weltraumschaf** [comment]
@the-simmon Can't rebase. You must rebase by your own.

**Pull request merged**